Studio support level of components/xblocks

cahrens opened the pull request:
### Description
 
[TNL-4670](https://openedx.atlassian.net/browse/TNL-4670)

Functionality:
If [XBlockStudioConfigurationFlag](https://cahrens.sandbox.edx.org/admin/xblock_django/xblockstudioconfigurationflag/) is enabled, support levels display next to Advanced Modules and HTML/Problem template types in the "new component" menus. A legend is present, including "Not Supported" only if the Advanced Setting to allow unsupported components is enabled (by default, this is disabled). "edX Support" link goes to Read the Docs (which is ugly because it is hardcoded in the template, but this is part of a much larger issue that is pervasive throughout the platform).

![image](https://cloud.githubusercontent.com/assets/484484/16927986/5012e552-4cfe-11e6-874e-a09413563453.png)

XBlocks disabled in [XBlockConfiguration](https://cahrens.sandbox.edx.org/admin/xblock_django/xblockconfiguration/) will be excluded from the "new component" menus. Other than that, all configuration is done through [XBlockStudioConfiguration](https://cahrens.sandbox.edx.org/admin/xblock_django/xblockstudioconfiguration/).

Note: to access admin django tables, use the account "admin" with password "edx".

### Sandbox
- [x] https://studio-cahrens.sandbox.edx.org/container/block-v1:edX+DemoX+Demo_Course+type@vertical+block@de3579b9ebc1461c8ac91b5d645e6628 (log in as staff or admin).

And a sandbox with the configuration flag disabled to verify that everything continues working as it currently does in production-- https://studio-flagoff.sandbox.edx.org.

### Testing
- [x] i18n
- [x] RTL
- [x] safecommit shows 0 violations.
- [x] Unit, integration, acceptance tests as appropriate

### Reviewers
If you've been tagged for review, please check your corresponding box once you've given the :+1:.
- [x] Code review: @staubina 
- [x] Code review: @efischer19 Thanks for your continued help! There's only one small deletion PR after this one.
- [x] Doc Review: @lamagnifica, @catong
- [x] UX review: @chris-mike 
- [x] Accessibility review: @clrux can you review just the underscore files included in the PR for a11y? Should be pretty fast (just the icons and support level legend).
- [x] Product review: @sstack22 

### Devops assistance
FYI @maxrothman. This has no DevOps dependencies, except that I will need to have the ansible playbook to set XBlockStudioConfiguration state run on prod/edge/staging before I merge this PR. I haven't created the DevOps tickets yet because I want to get sign-off on the support states first to minimize the number of times you guys have to run the playbook.

FYI: @cptvitamin

### Post-review
- [x] Squash commits

cahrens commented on common/lib/xmodule/xmodule/course_module.py:
```diff
@@ -830,6 +830,13 @@ class CourseFields(object):
         scope=Scope.settings
     )
+    allow_unsupported_xblocks = Boolean(
+        display_name=_("Allow the creation of unsupported XBlocks"),
```
@lamagnifica and @sstack22 I need help with wording here.

lamagnifica replied:
How about one of these? @sstack22  (Glad to explain my rationale behind any of the suggestions). (Please use initial caps.)
Allow Unsupported Problem Types
Allow Unsupported Problems and Tools
Allow Unsupported Course Content Types

sstack22 replied:
@lamagnifica  - I like the 2nd option "Allow Unsupported Problems and Tools".

cahrens replied:
Yes, note that this feature is not specific to problem types-- it controls HTML templates and XBlocks as well (really any xmodule/xblock that can be added through the "New" buttons).

cahrens replied:
@sstack22 and @lamagnifica How about "Allow the Creation of Unsupported Problems and Tools"?

The thing is that this setting only impacts whether or not you can create new ones. It doesn't impact whether existing unsupported components render/work. I'm afraid that without "Creation", it reads that you are toggling whether or not unsupported components work in the course.

lamagnifica replied:
OK, what about  "Add Unsupported Problems and Tools"?

cahrens replied:
OK, I'll go with "Add Unsupported Problems and Tools".

cahrens commented on cms/templates/js/add-xblock-component-menu.underscore:
```diff
@@ -10,20 +10,39 @@
         </ul>
-        <button class="cancel-button" data-type="<%= type %>"><%= gettext("Cancel") %></button>
+        <button class="cancel-button" data-type="<%- type %>"><%- gettext("Cancel") %></button>
+        <span class="support-documentation">
```
This will be moving into its own template, and I need to handle the following:

1. Don't show any of this of StudioXBlockConfigurationFlag is disabled
2. Pass in the platform name
3. Unfortunately are are stuck hard-coding the readthedocs link for now. :(

cahrens commented on cms/static/sass/elements/_modules.scss:
```diff
@@ -180,6 +180,37 @@
       }
     }
+
+    .support-documentation {
```
I have some additional styling tweaks to do, and I have to check if there are any responsive issues.

lamagnifica commented on common/lib/xmodule/xmodule/course_module.py:
```diff
@@ -830,6 +830,13 @@ class CourseFields(object):
+    allow_unsupported_xblocks = Boolean(
+        display_name=_("Allow the creation of unsupported XBlocks"),
+        help=_(
+            "If set to True, XBlocks that are not supported by edX can be created in Studio. Scary legal text..."
```
I suggest ( @sstack22 please weigh in too)

"Enter true or false. If true, additional, you can add unsupported content types to your course in Studio. Scary legal text."

(In the past, I have been told in no uncertain terms that we cannot include "edX" even for things that truly are edX-specific, such as the text on the video upload page, but up to you!)

cahrens replied:
We discussed this at the 4 PM meeting. The feeling is that we should go to this--

"Enter true or false. If true, you can add unsupported content types to your course in Studio. These content types are not recommended for use in courses due to non-compliance with one or more of the base requirements, such as testing, accessibility, internationalization, and documentation."

lamagnifica replied:
sweet.

cahrens commented on cms/djangoapps/contentstore/views/component.py:
```diff
@@ -47,17 +50,41 @@
-    disabled_create_block_types = XBlockDisableConfig.disabled_create_block_types()
-    return [c_type for c_type in ADVANCED_COMPONENT_TYPES if c_type not in disabled_create_block_types]
+    enabled_block_types = _filter_disabled_blocks(ADVANCED_COMPONENT_TYPES)
+    if XBlockStudioConfigurationFlag.is_enabled():
```
Note to reviewers: authorable_xblocks returns None if XBlockStudioConfigurationFlag.is_enabled() returns False. However, since there are other places where we explicitly check XBlockStudioConfigurationFlag.is_enabled(), I thought it was clearer to use that check here as well instead of comparing the return value of authorable_xblocks.

staubina replied:
Note, this comment is not a blocker, just a suggestion and question

Would it make sense to leave this check at this point and remove it from the component_support_level method?

This would allow for the component_support_level method to always return the currently set support level and the bypass would be handled in this central location.

The intent being that the component_support_level method would always return the real support level, which could be useful for reuse later.

cahrens replied:
@staubina Sure, this makes sense. I made the code change and also made it so that the Advanced Setting will not be displayed if XBlockStudioConfigurationFlat is false. This should enable me to merge this PR to master before all the individual xblock support states are worked out. https://github.com/edx/edx-platform/pull/12950/commits/2ba2f410c027a1945bf606a8f9ce03e7ac500e6f

cahrens commented on cms/djangoapps/contentstore/views/component.py:
```diff
@@ -189,57 +257,92 @@ def create_template_dict(name, cat, boilerplate_name=None, tab="common", hinted=
-        templates_for_category.append(create_template_dict(display_name, category, None, 'advanced'))
-        categories.add(category)
+
+        if support_level_without_template:
```
Note to reviewers: to easily see what changed here, add "?w=1" to the end of the github URL (ignore whitespace changes).

cahrens commented on cms/static/sass/_build-v1.scss:
```diff
@@ -10,7 +10,7 @@
 // +Base - Utilities
 // ====================
-@import 'variables';
+@import 'partials/variables';
```
Note: this was just a pre-existing typo. When I started using variables in _modules.scss, Lettuce tests failed due to "variables" not being properly included. I'm not sure why no failures occurred before, except that "variables" was only used in a couple of places with CMS (so perhaps no Lettuce tests hit the related files).

cahrens commented on cms/static/sass/elements/_modules.scss:
```diff
@@ -168,18 +168,49 @@
 
     // specific menu types
     &.new-component-problem {
-      padding-bottom: ($baseline/2);
```
Note to reviewers: did some adjustments so that the new component menu renders with the same spacing around the "Cancel" button and support legend for Advanced Modules, HTML, and Problems. Previously there were some vertical spacing differences.

cahrens commented on common/djangoapps/xblock_django/models.py:
```diff
@@ -41,27 +41,10 @@ def is_block_type_disabled(cls, block_type):
 
         return block_type in config.disabled_blocks.split()
 
-    @classmethod
```
Note to reviewers: this Django model will be completely deleted in the next story. I just can't remove it until everything is transitioned to the new models.

cahrens commented:
jenkins run python

cahrens commented:
@staubina and @efischer19 Please continue reviewing. I have made some small changes and refreshed the sandbox. There is also a sandbox linked in the description with the feature flag disabled, as I will likely merge this PR in that state while we continue to work out all the individual XBlock support states.

The acceptance test failure is an unrelated flaky test failure.

cahrens commented:
@sstack22 @chris-mike @lamagnifica Please review-- the sandbox is up-to-date and has the newest icons/messaging.

cahrens commented on cms/templates/js/add-xblock-component-support-legend.underscore:
```diff
@@ -0,0 +1,22 @@
+<% if (support_legend.show_legend) { %>
```
@clrux Please review for accessibility.

cahrens commented on cms/templates/js/add-xblock-component-support-level.underscore:
```diff
@@ -0,0 +1,10 @@
+<% if (support_level === "fs"){ %>
```
@clrux Please review for accessibility.

staubina commented:
@cahrens The update looks good to me.  

👍

efischer19 commented:
bokchoy failure might be flaky: `studio.test_studio_textbooks.TextbooksTest.test_create_first_book_message`

https://build.testeng.edx.org/job/edx-platform-bok-choy-pr/23167/

efischer19 commented:
jenkins run bokchoy

efischer19 commented on cms/djangoapps/contentstore/tests/test_course_settings.py:
```diff
@@ -784,6 +785,15 @@ def test_update_from_json_filtered_edxnotes_off(self):
+        """
+        self.assertNotIn('allow_unsupported_xblocks', CourseMetadata.fetch(self.fullcourse))
+        XBlockStudioConfigurationFlag(enabled=True).save()
+        self.assertIn('allow_unsupported_xblocks', CourseMetadata.fetch(self.fullcourse))
```
🏆  I love when tests are this easy to read

efischer19 commented:
👍  LGTM - very thorough!

chris-mike commented:
It looks like the fonts are not consistent in the Problem component.  The problem types and tabs seem to be using a font/style that's different than the legend's font/style (legend looks like open sans).  It looks like the fonts ARE consistent (problem types and legend) in the Advanced component.

Two suggestions:
1. Let's make all fonts the same - Opens Sans preferably, 16px (default size).
2. Can we make the legend/key one step smaller than the problem types?  So if problem types are open sans 16px, can we make the legend open sans 14px.

Related to the link and on-hover text, the colors (default and on-hover) do not pass WCAG AA.  I'd suggest the default link blue color is #0079BC (Use this color in your sass:palette(primary, base) rgb(0, 121, 188) #0079BC).  
On hover should be a darker blue (Use this color in your sass:palette(primary, dark) rgb(6, 86, 131) #065683) and should also be underlined on over.

cahrens commented:
jenkins run python

cahrens commented:
@chris-mike I have updated the sandbox with the following changes--

1) Changed font to Open Sans on the problem menu. That menu (because of the tabs) uses a JQuery component, and it was setting the font to something besides our standard. Thanks for noticing-- I certainly never did!

2) Changed font size of the legend to 14.

3) Updated the link and on-hover colors to the colors that @clrux made the link defaults in the recent PR #12953 (see _base.scss). These are not the colors that you recommended, but I think we should be consistent. Note that I have to explicitly define these colors instead of inheriting them because of the JQuery component that caused the #1 issue as well (I was seeing the default link colors for all of the menus except the problem one). I didn't underline on hover because we don't generally do that for links in _base.scss; let's chat if you want to change the default behavior.

cahrens commented on cms/static/sass/elements/_modules.scss:
```diff
@@ -168,18 +168,45 @@
       }
     }
+
+    .support-documentation {
```
@alisan617 Could you do a quick review of my sass changes? Unfortunately I'm having to override some things that I would have inherited from our base sass because we use a JQuery widget that defines its own styling.

cahrens commented:
@clrux and @chris-mike After talking with Chris, I have added in the underline to these links. I tried adding it in general in _base.scss as well, but it looked odd because there are too many things in Studio that are being marked up as links but should really be buttons. So for now, only adding the underline in this one place.

The sandbox is in the process of being updated....

cahrens commented:
jenkins run a11y

alisan617 commented on cms/static/sass/elements/_modules.scss:
```diff
@@ -168,18 +168,47 @@
+          text-decoration: underline;
+        }
+
+        padding-right: ($baseline/2);
```
it will be easier to read if padding-right is moved under text-decoration

cahrens replied:
Sure, done.

alisan617 commented on cms/static/sass/partials/_variables.scss:
```diff
@@ -39,6 +39,11 @@ $f-monospace: 'Bitstream Vera Sans Mono', Consolas, Courier, monospace;
 
+// +Colors - UXPL new pattern library colors
+// ====================
+$primary-accent: rgb(14, 166, 236);
```
Are these colors supposed to be the same as $uxpl-blue-base and $uxpl-blue-hover-active?

cahrens replied:
Good question-- I don't know! @clrux what should "primary-accent" be now? I was trying to define variables using the pattern library names to make it easier to understand how it links up with the pattern library.

I actually don't need primary-dark anymore and can remove it.

cahrens replied:
@alisan617 @clrux will update the colors as necessary after the pattern library changes are made. I have made the variable name match the style of the other uxpl colors that Chris recently submitted.

chris-mike commented:
All changes look good to me. Thanks!
👍

sstack22 commented:
@cahrens this looks great to me. 👍

cahrens commented:
@lamagnifica Do you wish to do a quick review?

lamagnifica commented on cms/djangoapps/contentstore/tests/test_course_settings.py:
```diff
@@ -784,6 +785,15 @@ def test_update_from_json_filtered_edxnotes_off(self):
+    def test_allow_unsupported_xblocks(self):
+        """
+        allow_unsupported_xblocks is only shown in Advanced Settings if
+        XBlockStudioConfiguartionFlag is enabled.
```
sp: Configuration?

lamagnifica commented on cms/djangoapps/contentstore/views/component.py:
```diff
@@ -152,24 +179,65 @@ def get_component_templates(courselike, library=False):
+            "show_legend": XBlockStudioConfigurationFlag.is_enabled(),
+            "allow_unsupported_xblocks": allow_unsupported,
+            "documentation_label": _("{platform_name} Support:").format(platform_name=settings.PLATFORM_NAME)
         }
```
:+

lamagnifica commented on cms/templates/js/add-xblock-component-support-legend.underscore:
```diff
@@ -0,0 +1,22 @@
+<% if (support_legend.show_legend) { %>
+    <span class="support-documentation">
+        <a class="support-documentation-link"
+           href="http://edx.readthedocs.io/projects/edx-partner-course-staff/en/latest/exercises_tools/create_exercises_and_tools.html#id3" target="_blank">
```
👍

lamagnifica commented on common/lib/xmodule/xmodule/course_module.py:
```diff
@@ -830,6 +830,15 @@ class CourseFields(object):
+        help=_(
+            "Enter true or false. If true, you can add unsupported content types to your course in Studio. "
+            "These content types are not recommended for use in courses due to non-compliance with one or "
+            "more of the base requirements, such as testing, accessibility, internationalization, and documentation."
```
👍

catong replied:
@cahrens @sstack22 Per a conversation we had yesterday we also need to update the help text under the Advanced Module policy to indicate that these types won't be available in your course unless you have specified a value of true  for "Add Unsupported Problems and Tools".

catong replied:
@cahrens @lamagnifica @sstack22 I'd like to suggest a couple of small tweaks to this Help text. Second sentence replace "These" with "Unsupported'.
Unless it has already been discussed and rejected, I'd also suggest replacing "content types" with "problems and tools", in line with terms we use in the title.
"...you can add unsupported problems and tools. ... Unsupported problems and tools are not recommended..."

lamagnifica commented:
user visible strings look good to me, thanks 👍

clrux commented on cms/templates/js/add-xblock-component-menu-problem.underscore:
```diff
@@ -19,15 +19,17 @@
-                        <button type="button" class="button-component" data-category="<%= templates[i].category %>">
-                            <span class="name"><%= templates[i].display_name %></span>
+                        <button type="button" class="button-component" data-category="<%- templates[i].category %>">
+                            <%= HtmlUtils.HTML(support_indicator_template({support_level: templates[i].support_level})) %>
```
Do you need the `<%-` here instead of the `=`?

cahrens replied:
No-- that would escape the template so that users would visually see the HTML tags. You use <%- for plain text that should be escaped.

http://edx.readthedocs.io/projects/edx-developer-guide/en/latest/conventions/safe_templates.html#underscore-js-template-files

If you need to pass an HTML snippet to a template, which has already been HTML-escaped, you should name the variable with an Html suffix, and use HtmlUtils.ensureHtml() to ensure that it was in fact properly HTML- escaped. See the following example.

````
<%= HtmlUtils.ensureHtml(nameHtml) %>

clrux commented on cms/templates/js/add-xblock-component-menu-problem.underscore:
```diff
@@ -19,15 +19,17 @@
-                            <span class="name"><%= templates[i].display_name %></span>
+                        <button type="button" class="button-component" data-category="<%- templates[i].category %>"
+                           data-boilerplate="<%- templates[i].boilerplate_name %>">
+                            <%= HtmlUtils.HTML(support_indicator_template({support_level: templates[i].support_level})) %>
```
Do you need the `<%-` here instead of the `=`?

clrux commented on cms/templates/js/add-xblock-component-menu-problem.underscore:
```diff
@@ -40,14 +42,16 @@
-                        <span class="name"><%= templates[i].display_name %></span>
+                    <button type="button" class="button-component" data-category="<%- templates[i].category %>"
+                       data-boilerplate="<%- templates[i].boilerplate_name %>">
+                        <%= HtmlUtils.HTML(support_indicator_template({support_level: templates[i].support_level})) %>
```
Do you need the `<%-` here instead of the `=`?

clrux commented on cms/templates/js/add-xblock-component-menu-problem.underscore:
```diff
@@ -40,14 +42,16 @@
     </div>
-    <button class="cancel-button" data-type="<%= type %>"><%= gettext("Cancel") %></button>
+    <button class="cancel-button" data-type="<%- type %>"><%- gettext("Cancel") %></button>
+    <%= HtmlUtils.HTML(support_legend_template({support_legend: support_legend})) %>
```
Do you need the `<%-` here instead of the `=`?

clrux commented on cms/templates/js/add-xblock-component-menu.underscore:
```diff
@@ -10,20 +10,23 @@
-                    <button type="button" class="button-component" data-category="<%= templates[i].category %>">
-                        <span class="name"><%= templates[i].display_name %></span>
+                    <button type="button" class="button-component" data-category="<%- templates[i].category %>">
+                        <%= HtmlUtils.HTML(support_indicator_template({support_level: templates[i].support_level})) %>
```
Do you need the `<%-` here instead of the `=`?

clrux commented on cms/templates/js/add-xblock-component-menu.underscore:
```diff
@@ -10,20 +10,23 @@
-                        <span class="name"><%= templates[i].display_name %></span>
+                    <button type="button" class="button-component" data-category="<%- templates[i].category %>"
+                       data-boilerplate="<%- templates[i].boilerplate_name %>">
+                        <%= HtmlUtils.HTML(support_indicator_template({support_level: templates[i].support_level})) %>
```
Do you need the `<%-` here instead of the `=`?

clrux commented on cms/templates/js/add-xblock-component-menu.underscore:
```diff
@@ -10,20 +10,23 @@
         </ul>
-        <button class="cancel-button" data-type="<%= type %>"><%= gettext("Cancel") %></button>
+        <button class="cancel-button" data-type="<%- type %>"><%- gettext("Cancel") %></button>
+        <%= HtmlUtils.HTML(support_legend_template({support_legend: support_legend})) %>
```
Do you need the `<%-` here instead of the `=`?

clrux commented:
@cahrens Looks good. I had a few questions surrounding safe templates. I feel confident that you know what you're doing, but thought I'd comment just in case. I did find one issue:

The "cahrens Support" link in the legend isn't really descriptive enough. I understand that this link might say "<insert name here> Support", but I think we need to add some text that clarifies what this legend is about. Maybe something like "Information on supported, provisionally supported, and unsupported components...". That's wordy, but perhaps @lamagnifica can smith those words?

lamagnifica commented:
@clrux would "{platform_name} Support Category" work? Will screen readers not keep the "Supported" "Provisional" and "Not Supported" labels for the icons together with this group label?

cahrens commented:
@lamagnifica how about "{platform_name} Support Categories:"?

I don't think that screen readers need the legend beyond accessing the link to the documentation. The icons all have sr text associated with them, so they don't need to map icons to text.

cahrens commented:
@lamagnifica Actually, do you like "{platform_name} Support Levels:" better than "Categories"? And I assume you do want the colon at the end, but let me know if that isn't true.

lamagnifica commented:
before we tweak my suggestion too much, @cahrens, let's figure out why @clrux suggested such a long label for the three category/level names...!

clrux commented:
@lamagnifica I wasn't suggesting that label exactly :) I was suggesting it as context for what the link does or why the link/support might be needed. The idea here is to convey that "a user can get more information on what each of these categories mean". General "platform support" sounds daunting. Make sense?

lamagnifica commented:
@clrux and I chatted offline. 
The icon descriptions are read for the list items, but not st the bottom of the list
Must have: context for where the link will take you, including when the screen reader "link list" option is used (it reads all links on the page in a list, without any other context for any of them)
How about one of these suggestions? 
1st set: (these follow the template used in the right info pane throughout Studio)
Learn more about the {platform_name} support levels
Learn more about the levels of support
Learn more about the support levels
Learn more about support levels

2nd set: (more descriptive context without using the Learn More pattern)
About the {platform_name} Support Levels
Help for Support Levels
What are the support levels?
Levels of Support Definitions
Levels of Support Defined

@sstack22 , @catong

cahrens commented:
I advocate for making whatever is visible as short as possible. We can add additional screen-reader only text if necessary.

Having talked to @clrux myself after @lamagnifica did, I think that "edX Support Levels" suffices both for visual and screen readers.

catong commented:
@cahrens Agree with "edX Support Levels"

cahrens commented:
@catong The sandbox is updated with the doc changes-- please review.

catong commented:
@cahrens was it an explicit decision for the doc link to be http://edx.readthedocs.io/projects/edx-partner-course-staff/en/latest/exercises_tools/create_exercises_and_tools.html#id3?

This actually links to a heading within a list of links, rather than to the Support Level topic itself, and results in "Levels of Support" heading appearing down the browser page. 

I think a better link would be:
http://edx.readthedocs.io/projects/edx-partner-course-staff/en/latest/exercises_tools/create_exercises_and_tools.html#levels-of-support-for-tools

cahrens commented:
Not an explicit decision-- just my mistake. I'll fix it and squash at the same time.

catong commented:
@cahrens Can I make one more tweak to the help text for the Advanced Modules advanced setting:
"will not be available" -> "are not available"

catong commented:
👍 from me with the final 2 fixes: doc target and tweak to the Advanced Modules help text.

cahrens commented:
@catong Sandbox is updated again-- will merge once tests pass.

cahrens commented:
Actually, I just realized that we can't reference "Add Unsupported Problems and Tools" from the "Advanced Module List" help text. We only show "Add Unsupported Problems and Tools" when the feature flag is enabled, and for most open source installations, they will never enable it (as it requires explicitly listing every xblock/template combination).

I will be removing that help text momentarily....

cahrens commented:
jenkins run python

cahrens commented:
Sandboxes are updated, including the one with the feature flag off (which is how this will be initially submitted).

https://studio-flagoff.sandbox.edx.org/course/course-v1:edX+DemoX+Demo_Course
